### ISSUE TITLE
Libgpgme.so hard coded path leads to build failers for pkgs that depend on it.

### ISSUE THREAD

**Issue by khallebal** [haikuports/haikuports#2093]
The libgpgme.so symlink is not part of the standard package it's only part of the devel package as it is the case for many packages (is this intentionel BTW?), so makefile seems to look  for it only in /system/lib dir it does not "bother" to look in the /system/develop/lib where it resides, cmake seems to set during configure only /boot/system/lib as a the path for libs to link against.
How should it be fixed?

this is the error i'm getting while compiling the download manager kget.

```
`*** No rule to make target '/packages/gpgme_x86-1.9.0-2/.self/lib/x86/libgpgme.so', needed by 'libkgetcore.so.5.0.0'.  Stop.
`
```

**Comment by diversys** [comment]
`libgpgme.so.11.18.0` is part of the base package.

```
package list /system/packages/gpgme_x86-1.9.0-2-x86_gcc2.hpkg | grep libgpgme.so
    libgpgme.so.11                       0  2017-11-12 01:01:36  lrwxrwxrwx  -> libgpgme.so.11.18.0
    libgpgme.so.11.18.0            1246908  2017-11-12 01:01:36  -rwxr-xr-x

```

**Comment by pulkomandy** [comment]
On Tue, Jan 09, 2018 at 11:21:09PM +0000, khallebal wrote:
> The libgpgme.so symlink is not part of the standard package it's only part of the devel package as it is the case for many packages (is this intentionel BTW?), so makefile seems to look  for it only in /system/lib dir it does not "bother" to look in the /system/develop/lib where it resides, cmake seems to set during configure only /boot/system/lib as a the path for libs to link against.
> How should it be fixed?

This is intentional, and the build system for your program should be
fixed. There is probably a "findgpgme.cmake" or similar, either inside
cmake itself or as part of the buildsystem of kget.

Background information:
Libraries can come in different versions as people continue developping
them. Different versions may be incompatible. To avoid confusion, a
version is appended to the filename, and at runtime, applications should
only use that versionned name to refer to the library. This way,
applications needing different versions of the library can exist side by
side.

However, when compiling a program, you will use linker flags such as
'-lgpgme'. Notice that this does not include any version information.
The idea is that in that case, you always want to link with the most
recent version of the library.

In Haiku we tried to make this distinction more clear. So, the main
application package has only the versionned library, the one you need to
run the program. It cannot be used alone to link with something. The
_devel package contains everything you need to compile softwware that
links against the library. And to make things even clearer, all the
development files are in /system/develop.

This should not be a problem, as long as buildsystems are well-behaving
and use pkg-config or similar tools to locate libraries. Unfortunately,
there are some old cmake modules (these find*.cmake files I was talking
about earlier), which just use a few hardcoded paths to look for
libraries. The good news is that people are usually quite happy to
accept changes to these files, so that pkg-config is used (modern cmake
versions provide the required support for that).

**Comment by khallebal** [comment]
Ok, so it seems there is nothing that can be done on the cmake side for this particular case, there is only one thing left to do and that is the use of CFLAGS/CXXFLAGS to tell cmake to set the devel path as an extra location to look for the libs? i am not very familiar with cmake, could someone tell me how to use CFLAGS/CXXFLAGS in with cmake? and close this ticket
Bye.

**Comment by pulkomandy** [comment]
On Wed, Jan 10, 2018 at 10:47:32PM +0000, khallebal wrote:
> Ok, so it seems there is nothing that can be done on the cmake side for this particular case, there is only one thing left to do and that is the use of CFLAGS/CXXFLAGS to tell cmake to set the devel path as an extra location to look for the libs? i am not very familiar with cmake, could someone tell me how to use CFLAGS/CXXFLAGS in with cmake? and close this ticket

As I wrote above, this is all driven by "modules" in cmake. Somewhere,
either in the cmake distribution or in the project you are compiling,
there is a "FindGpgme.cmake" file. This file should be fixed to use
pkg-config, instead of using hardcoded paths. This is the right way to
fix the problem.

**Comment by khallebal** [comment]
Well, it starts to look more like a bug somewhere, i just can't put my finger on it.
Bear with me please.
The project i am compiling does absolutely not need libgpgme, it only needs the ++ and the Qt versions of it (i.e, gpgmepp and Qgpgme), to make sure i'm not mistaking, i grepped the source for the word "gpgme" before and after calling Cmake, there was no trace of gpgme in any file, so i am puzzled as to why gcc tries to link against it.
 Thoughts?.

**Comment by pulkomandy** [comment]
Ah, I see.

This can happen if one of the libraries you use itself depends on another library which is in another package. This should be sorted by pkg-config files, but it's possible that one of the recipes forgot to do the "fixPkgConfig" to make sure the files themselves point to the right place. Or, maybe the pkg-config files are not found, and cmake will itself try to guess something.

Another possibility is stray .la (libtool) files leading to extra linking, which isn't needed on Haiku. You can check your lib and develop/lib directories and make sure there are no .la files leftover there. If there are, they should be removed from packages.

**Comment by khallebal** [comment]
I believe pkg-config is unrelated in this case because Cmak uses "GpgpmeppConfig.cmake" and "QGpgmeConfig.cmake" to find respectively libQgpgme and libgpgmepp, so if there is an error it's probably in one of these two files which are automatically generated by Cmake . I 'll paste the contents of them here maybe someone can spot the error.

GpgpmeppConfig.cmake:
-------------------------------------------------------------------------
```
# CMake Config file for GPGMEPP.
# Copyright (C) 2016 Intevation GmbH
#
# This file is part of GPGMEPP.
#
# GPGME-CL is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# GPGME-CL is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA

# based on a generated file from cmake.
# Generated by CMake 3.0.2

if("${CMAKE_MAJOR_VERSION}.${CMAKE_MINOR_VERSION}" LESS 2.5)
   message(FATAL_ERROR "CMake >= 2.6.0 required")
endif()
cmake_policy(PUSH)
cmake_policy(VERSION 2.6)
#----------------------------------------------------------------
# Generated CMake target import file.
#----------------------------------------------------------------

# Commands may need to know the format version.
set(CMAKE_IMPORT_FILE_VERSION 1)

# Protect against multiple inclusion, which would fail when already imported targets are added once more.
set(_targetsDefined)
set(_targetsNotDefined)
set(_expectedTargets)
foreach(_expectedTarget Gpgmepp)
  list(APPEND _expectedTargets ${_expectedTarget})
  if(NOT TARGET ${_expectedTarget})
    list(APPEND _targetsNotDefined ${_expectedTarget})
  endif()
  if(TARGET ${_expectedTarget})
    list(APPEND _targetsDefined ${_expectedTarget})
  endif()
endforeach()
if("${_targetsDefined}" STREQUAL "${_expectedTargets}")
  set(CMAKE_IMPORT_FILE_VERSION)
  cmake_policy(POP)
  return()
endif()
if(NOT "${_targetsDefined}" STREQUAL "")
  message(FATAL_ERROR "Some (but not all) targets in this export set were already defined.\nTargets Defined: ${_targetsDefined}\nTargets not yet defined: ${_targetsNotDefined}\n")
endif()
unset(_targetsDefined)
unset(_targetsNotDefined)
unset(_expectedTargets)

# Create imported target Gpgmepp
add_library(Gpgmepp SHARED IMPORTED)

set_target_properties(Gpgmepp PROPERTIES
  INTERFACE_INCLUDE_DIRECTORIES "/packages/gpgme_x86-1.9.0-2/.self/develop/headers/x86/gpgme++;/packages/gpgme_x86-1.9.0-2/.self/develop/headers/x86"
  INTERFACE_LINK_LIBRARIES "pthread;/packages/gpgme_x86-1.9.0-2/.self/lib/x86/libgpgme.so;-L/packages/libassuan_x86-2.4.4-1/.self/lib/x86 -lassuan -lnetwork -L/packages/libassuan_x86_devel-2.4.4-1/devel~libgpg_error_x86/develop/lib/x86 -lgpg-error"
  IMPORTED_LOCATION "/packages/gpgme_x86-1.9.0-2/.self/lib/x86/libgpgmepp.so"
)

if(CMAKE_VERSION VERSION_LESS 2.8.12)
  message(FATAL_ERROR "This file relies on consumers using CMake 2.8.12 or greater.")
endif()

# Loop over all imported files and verify that they actually exist
foreach(target ${_IMPORT_CHECK_TARGETS} )
  foreach(file ${_IMPORT_CHECK_FILES_FOR_${target}} )
    if(NOT EXISTS "${file}" )
      message(FATAL_ERROR "The imported target \"${target}\" references the file
   \"${file}\"
but this file does not exist.  Possible reasons include:
* The file was deleted, renamed, or moved to another location.
* An install or uninstall procedure did not complete successfully.
* The installation package was faulty and contained
   \"${CMAKE_CURRENT_LIST_FILE}\"
but not all the files it references.
")
    endif()
  endforeach()
  unset(_IMPORT_CHECK_FILES_FOR_${target})
endforeach()
unset(_IMPORT_CHECK_TARGETS)

# Commands beyond this point should not need to know the version.
set(CMAKE_IMPORT_FILE_VERSION)
cmake_policy(POP)

get_filename_component(QGpgme_DIR "${CMAKE_CURRENT_LIST_FILE}" PATH)
# Pull in QGpgme for compatibility with KF5 variant.
find_package(QGpgme CONFIG)

```
--------------------------------------------------------------------------
QGpgmeConfig.cmake
---------------------------------------------------------------------------
```
# CMake Config file for QGPGME.
# Copyright (C) 2016 Intevation GmbH
#
# This file is part of GPGME.
#
# GPGME-CL is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# GPGME-CL is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA

# based on a generated file from cmake.
# Generated by CMake 3.0.2

if("${CMAKE_MAJOR_VERSION}.${CMAKE_MINOR_VERSION}" LESS 2.5)
   message(FATAL_ERROR "CMake >= 2.6.0 required")
endif()
cmake_policy(PUSH)
cmake_policy(VERSION 2.6)
#----------------------------------------------------------------
# Generated CMake target import file.
#----------------------------------------------------------------

# Commands may need to know the format version.
set(CMAKE_IMPORT_FILE_VERSION 1)

# Protect against multiple inclusion, which would fail when already imported targets are added once more.
set(_targetsDefined)
set(_targetsNotDefined)
set(_expectedTargets)
foreach(_expectedTarget QGpgme)
  list(APPEND _expectedTargets ${_expectedTarget})
  if(NOT TARGET ${_expectedTarget})
    list(APPEND _targetsNotDefined ${_expectedTarget})
  endif()
  if(TARGET ${_expectedTarget})
    list(APPEND _targetsDefined ${_expectedTarget})
  endif()
endforeach()
if("${_targetsDefined}" STREQUAL "${_expectedTargets}")
  set(CMAKE_IMPORT_FILE_VERSION)
  cmake_policy(POP)
  return()
endif()
if(NOT "${_targetsDefined}" STREQUAL "")
  message(FATAL_ERROR "Some (but not all) targets in this export set were already defined.\nTargets Defined: ${_targetsDefined}\nTargets not yet defined: ${_targetsNotDefined}\n")
endif()
unset(_targetsDefined)
unset(_targetsNotDefined)
unset(_expectedTargets)

# Create imported target QGpgme
add_library(QGpgme SHARED IMPORTED)

set_target_properties(QGpgme PROPERTIES
  INTERFACE_INCLUDE_DIRECTORIES "/packages/gpgme_x86-1.9.0-2/.self/develop/headers/x86/qgpgme;/packages/gpgme_x86-1.9.0-2/.self/develop/headers/x86"
  INTERFACE_LINK_LIBRARIES "Gpgmepp;Qt5::Core"
  IMPORTED_LOCATION "/packages/gpgme_x86-1.9.0-2/.self/lib/x86/libqgpgme.so"
)

if(CMAKE_VERSION VERSION_LESS 2.8.12)
  message(FATAL_ERROR "This file relies on consumers using CMake 2.8.12 or greater.")
endif()

# Loop over all imported files and verify that they actually exist
foreach(target ${_IMPORT_CHECK_TARGETS} )
  foreach(file ${_IMPORT_CHECK_FILES_FOR_${target}} )
    if(NOT EXISTS "${file}" )
      message(FATAL_ERROR "The imported target \"${target}\" references the file
   \"${file}\"
but this file does not exist.  Possible reasons include:
* The file was deleted, renamed, or moved to another location.
* An install or uninstall procedure did not complete successfully.
* The installation package was faulty and contained
   \"${CMAKE_CURRENT_LIST_FILE}\"
but not all the files it references.
")
    endif()
  endforeach()
  unset(_IMPORT_CHECK_FILES_FOR_${target})
endforeach()
unset(_IMPORT_CHECK_TARGETS)

# This file does not depend on other imported targets which have
# been exported from the same project but in a separate export set.

# Commands beyond this point should not need to know the version.
set(CMAKE_IMPORT_FILE_VERSION)
cmake_policy(POP)

# Gpgmepp is a requirement for qgpgme
find_package(Gpgmepp CONFIG REQUIRED)

```

**Comment by phoudoin** [comment]
The GpgpmeppConfig.cmake file is generated, and if look at the GpgpmeppConfig.cmake.in.in template file the INTERFACE_LINK_LIBRARIES value is built from:

`"pthread;@resolved_libdir@/libgpgme@libsuffix@;@LIBASSUAN_LIBS@"`

Which resolved into hardcoding at build time location of Gpgmepp's dependency on libgpgme.so, by:
1) using an absolute path to package .self/lib/... (ugly)
2) asserting the unversionned shared library/symlink will be in the same place. (linuxy)

Which break on Haiku, due to runtime vs devel packages distinct layout.
I guess patching this file to find libgpgme.so dynamically when used on Haiku could do it:

> find_library(LIBGPGME gpgme)
> set_target_properties(Gpgmepp PROPERTIES
>     INTERFACE_INCLUDE_DIRECTORIES "@resolved_includedir@/gpgme++;@resolved_includedir@"
>     INTERFACE_LINK_LIBRARIES "pthread;${LIBGPGME};@LIBASSUAN_LIBS@"
>     IMPORTED_LOCATION "@resolved_libdir@/libgpgmepp@libsuffix@"
> )

**Comment by fbrosson** [comment]
If I understand correctly, you are working on a recipe for kget.

Did you forget to mention `devel:libassuan$secondaryArchSuffix` in `BUILD_REQUIRES` or `cmd:pkg_config$secondaryArchSuffix` in `BUILD_PREREQUIRES` ?

(If you got both of these then I don't know why you are getting the `@LIBASSUAN_LIBS@`.)

**Comment by phoudoin** [comment]
It's a placeholder  in the template file.
The generated one built by gpgme package recipe have path pointing at when the libassuan package places it's libs.

No the main issue is that the libgpgme.so path is still too much hardcoded in the template. Fixing this will fix the issue.

Which has nothing to do with cmake not looking in /system/develop/lib btw as it does ... But only if you use it instead of enforcing hardcoded paths.

**Comment by fbrosson** [comment]
OK, thanks. I got it :)

**Comment by khallebal** [comment]
> If I understand correctly, you are working on a recipe for kget.

yes.


> Did you forget to mention devel:libassuan$secondaryArchSuffix in BUILD_REQUIRES or cmd:pkg_config$secondaryArchSuffix in BUILD_PREREQUIRES ?

I have cmd:pkg_config$secondaryArchSuffix set in BUILD_PREREQUIRES, but not 
devel:libassuan$secondaryArchSuffix in BUILD_REQUIRES, devel:libassuan is not a kget dependency, but a gpgme pkg dependency, so maybe the solution as pointed out by phoudoin is to patch the GpgpmeppConfig.cmake.in.in template in the gpgme pkg?

> Which has nothing to do with cmake on Haiku not looking in /system/develop/lib btw.

Yeah i realise the title is misleading, that's because i didn't know what was going on exactly before you pointed out the real problem.

**Comment by korli** [comment]
fixed in cc90d358cb0341e8d5ac42db21753d6f5365d19c
